File: app/src/main/java/com/zionhuang/music/playback/MediaSessionConnection.java
package com.zionhuang.music.playback;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.IBinder;
import android.os.RemoteException;
import android.support.v4.media.MediaMetadataCompat;
import android.support.v4.media.session.MediaControllerCompat;
import android.support.v4.media.session.MediaSessionCompat;
import android.support.v4.media.session.PlaybackStateCompat;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.google.android.exoplayer2.ui.PlayerView;

import java.util.List;

public class MediaSessionConnection {
    private Context mContext;
    private PlayerView mPlayerView;
    private MediaControllerCompat mMediaController;
    private MediaControllerCallback mMediaControllerCallback = new MediaControllerCallback();
    private MediaServiceConnection mServiceConnection;
    private MutableLiveData<Boolean> isConnected = new MutableLiveData<>();
    private MutableLiveData<PlaybackStateCompat> playbackState = new MutableLiveData<>();
    private MutableLiveData<MediaMetadataCompat> nowPlaying = new MutableLiveData<>();

    public MediaSessionConnection(Context context) {
        mContext = context;
    }

    public void connect() {
        Intent intent = new Intent(mContext, MusicService.class);
        mServiceConnection = new MediaServiceConnection();
        mContext.bindService(intent, mServiceConnection, Context.BIND_AUTO_CREATE);
    }

    public void disconnect() {
        mContext.unbindService(mServiceConnection);
    }

    public void setPlayerView(PlayerView playerView) {
        mPlayerView = playerView;
    }

    @NonNull
    public LiveData<Boolean> getIsConnected() {
        return isConnected;
    }

    public LiveData<PlaybackStateCompat> getPlaybackState() {
        return playbackState;
    }

    public LiveData<MediaMetadataCompat> getNowPlaying() {
        return nowPlaying;
    }

    public MediaControllerCompat getMediaController() {
        return mMediaController;
    }

    @Nullable
    public MediaControllerCompat.TransportControls getTransportControls() {
        return mMediaController != null ? mMediaController.getTransportControls() : null;
    }

    private class MediaControllerCallback extends MediaControllerCompat.Callback {
        @Override
        public void onPlaybackStateChanged(PlaybackStateCompat state) {
            playbackState.postValue(state);
        }

        @Override
        public void onMetadataChanged(MediaMetadataCompat metadata) {
            nowPlaying.postValue(metadata);
        }

        @Override
        public void onQueueChanged(List<MediaSessionCompat.QueueItem> queue) {
            super.onQueueChanged(queue);
        }

        @Override
        public void onSessionDestroyed() {
            super.onSessionDestroyed();
        }
    }

    private class MediaServiceConnection implements ServiceConnection {

        @Override
        public void onServiceConnected(ComponentName name, IBinder service) {
            MusicService mService = ((MusicService.LocalBinder) service).getService();
            mService.setPlayerView(mPlayerView);
            try {
                mMediaController = new MediaControllerCompat(mContext, mService.getSessionToken());
                mMediaController.registerCallback(mMediaControllerCallback);
                isConnected.postValue(true);
            } catch (RemoteException e) {
                e.printStackTrace();
            }
        }

        @Override
        public void onServiceDisconnected(ComponentName name) {
            mMediaController.unregisterCallback(mMediaControllerCallback);
            isConnected.postValue(false);
        }
    }
}

File: app/src/main/java/com/zionhuang/music/playback/MediaSessionConnection.kt
package com.zionhuang.music.playback

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.os.IBinder
import android.os.RemoteException
import android.support.v4.media.MediaMetadataCompat
import android.support.v4.media.session.MediaControllerCompat
import android.support.v4.media.session.MediaSessionCompat.QueueItem
import android.support.v4.media.session.PlaybackStateCompat
import com.zionhuang.music.playback.MusicService.MusicBinder
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow

object MediaSessionConnection {
    var mediaController: MediaControllerCompat? = null
        private set
    val transportControls: MediaControllerCompat.TransportControls? get() = mediaController?.transportControls
    private val mediaControllerCallback = MediaControllerCallback()

    private val _isConnected = MutableStateFlow(false)
    private val _playbackState = MutableStateFlow<PlaybackStateCompat?>(null)
    private val _mediaMetadata = MutableStateFlow<MediaMetadataCompat?>(null)
    private val _queueTitle = MutableStateFlow<String?>(null)
    private val _queueItems = MutableStateFlow<List<QueueItem>>(emptyList())

    val isConnected: StateFlow<Boolean> = _isConnected
    val playbackState: StateFlow<PlaybackStateCompat?> = _playbackState
    val mediaMetadata: StateFlow<MediaMetadataCompat?> = _mediaMetadata
    val queueTitle: StateFlow<String?> = _queueTitle
    val queueItems: StateFlow<List<QueueItem>> = _queueItems

    private var _binder: MusicBinder? = null
    val binder: MusicBinder? get() = _binder

    private var serviceConnection: ServiceConnection? = null

    fun connect(context: Context) {
        serviceConnection = object : ServiceConnection {
            override fun onServiceConnected(name: ComponentName, iBinder: IBinder) {
                if (iBinder !is MusicBinder) return
                _binder = iBinder
                try {
                    mediaController = MediaControllerCompat(context, iBinder.sessionToken).apply {
                        registerCallback(mediaControllerCallback)
                    }
                    _isConnected.value = true
                } catch (_: RemoteException) {
                }
            }

            override fun onServiceDisconnected(name: ComponentName) {
                _binder = null
                mediaController?.unregisterCallback(mediaControllerCallback)
                _isConnected.value = false
            }
        }.also {
            val intent = Intent(context, MusicService::class.java)
            context.bindService(intent, it, Context.BIND_AUTO_CREATE)
        }
    }

    fun disconnect(context: Context) {
        if (serviceConnection != null) {
            context.unbindService(serviceConnection!!)
        }
    }

    private class MediaControllerCallback : MediaControllerCompat.Callback() {
        override fun onPlaybackStateChanged(state: PlaybackStateCompat) {
            _playbackState.value = state
        }

        override fun onMetadataChanged(metadata: MediaMetadataCompat?) {
            _mediaMetadata.value = metadata
            // force update playback state
            mediaController?.let {
                _playbackState.value = it.playbackState
            }
        }

        override fun onQueueChanged(queue: List<QueueItem>) {
            _queueItems.value = queue
        }

        override fun onQueueTitleChanged(title: CharSequence?) {
            _queueTitle.value = title?.toString()
        }
    }
}